app/queries/developers/query_item_builder.rb

```diff
@@ -0,0 +1,35 @@
+class Developers::QueryItemBuilder
+
+  attr_reader :query_items, :options
+
+  QUERY_ITEM_OPTIONS = %i[sort countries]
+
+  def initialize(options = {})
+    @options = options
+    @query_items = []
+  end
+
+  def build_query_items
+    add_sorting_query
+    add_countries_filter_query
+
+    query_items
+  end
+
+  def add_sorting_query
+    query_items << Developers::SortQueryItem.new(options[:sort])
+  end
+
+  def add_countries_filter_query
+    query_items << Developers::CountriesQueryItem.new(options[:countries])
+  end
+
+  # defines query item methods like sort_query_item, countries_query_item
+  # which can give us the respective query item object
+  # example: query_item_builder.sort_query_item
+  QUERY_ITEM_OPTIONS.each do |method_initial, _|
+    define_method "#{method_initial}_query_item" do
+      query_items.detect { |query_item| query_item.type.has_key? method_initial }
+    end
+  end
```

Comment:
This feels like it is breaking the point of the refactor. If the public interface exposes the different filters by name then anyone using the query needs to know about reach filter.

Reply:
@joemasilotti since methods like `sort`, `countries` etc are called on developer query object in many places (example: https://github.com/joemasilotti/railsdevs.com/pull/821/files#diff-9b36d228956c34411c77b9a8f1fc0f9831f453e3c4fb52b050673af85a6ee584R14) , somehow we have to expose those either in developer query or query builder. 

Any thoughts on how to achieve that if we don't get access to `sort_query_item` or `countries_query_item` ?

Reply:
One approach is for the query to expose an array of query items publicly. Then, the view can iterate over those to generate the necessary form elements.

The query item can indicate what type of field it requires (text box, select, checkbox, etc.) along with the copy and notes needed for the UI. It can also expose a method that decides if a collapse should be expanded or not.

Does that make sense?

Reply:
> One approach is for the query to expose an array of query items publicly. Then, the view can iterate over those to generate the necessary form elements.
> 
@joemasilotti made adjustments to countries filter and sort query items, let me know if this is fine.

 
> The query item can indicate what type of field it requires (text box, select, checkbox, etc.) along with the copy and notes needed for the UI. It can also expose a method that decides if a collapse should be expanded or not.
> 
@joemasilotti can you please explain a bit more what do you mean by indicating tags like checkbox, select, etc ?
Should we let the query item return its own html  and them embed it to `app/components/developers/query_component.html.erb` ?

Reply:
Thanks for the update!

In an ideal world we could pass in a bunch of `QueryItem` duck types and a single `SortItem` into `DeveloperQuery` and that would be that. The HTML could be generated based entirely by iterating over these items and inferring the type.

But the more I think about it the more I'm wondering if a) I'm overcomplicating it and b) we are rebuilding ransack.

I appreciate the work you put into this PR. But I think I'd like to explore ransack next. I'd still love to award a bounty - [send me an email](mailto:joe@railsdevs.com) and we can figure it out.